### PR TITLE
[release/6.0-staging] Allow setting ZipArchiveEntry general-purpose flag bits

### PR DESCRIPTION
**Pull request by carlossanlop**
Backport of #98278 to release/6.0-staging

/cc @carlossanlop @edwardneal

## Customer Impact

- [x] Customer reported
- [ ] Found internally

Originally reported in dotnet/Open-XML-SDK.

.NET Framework allows consumers of `System.IO.Packaging.ZipPackage` to use `CompressionOption` to update bits 1 & 2 in the ZIP archive header's general purpose bitfield. This functionality was lost in .NET Core.

OpenXML is affected because when a document like XSLX is created with our APIs, it is not compliant with the OpenXML specification ISO/IEC 29500-2, as the standard bits 1 and 2 are mandatory and are not reflecting the applied zip compression. Full discussions in https://github.com/dotnet/runtime/issues/88812 and  https://github.com/dotnet/Open-XML-SDK/issues/1443

Specifying the `CompressionLevel` of `System.IO.Compression.ZipArchive` will now change these bits, and the `ZipArchive`'s `CompressionLevel` value will be read from the bitfield.

## Regression

- [x] Yes
- [ ] No

This PR also introduces a breaking change which is reported in detail here: https://github.com/dotnet/docs/issues/40299 . Summarized, the `CompressionOption` values from `System.IO.Packaging` will map to different `CompressionLevel` levels in `System.IO.Compression` depending if using .NET Framework or .NET Core.

## Testing

New tests were added to both `System.IO.Compression` and `System.IO.Packaging` to verify that setting the `CompressionLevel` sets the expected bits, that they can be re-read correctly and that the file data can still be decompressed correctly.

## Risk

Low.

Thanks to @edwardneal for fixing the issue and @maedula for the detailed report of this problem.

### CODE REVIEW

File: src/libraries/System.IO.Packaging/tests/ReflectionTests.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                using StreamWriter sw = new(partStream);
+                sw.Write(Tests.s_DocumentXml);
+            }
+            package.CreateRelationship(part.Uri, TargetMode.Internal, "http://packageRelType", "rId1234");
+        }
+
+        ms.Position = 0;
+        using (ZipArchive archive = new ZipArchive(ms, ZipArchiveMode.Read, leaveOpen: false))
+        {
+            foreach (ZipArchiveEntry entry in archive.Entries)
```
Comment (edwardneal) [inline src/libraries/System.IO.Packaging/tests/ReflectionTests.cs:32]:
```suggestion
            FieldInfo archiveEncodingFieldInfo = typeof(ZipArchive).GetField("_entryNameAndCommentEncoding", BindingFlags.Instance | BindingFlags.NonPublic);
            System.Text.Encoding archiveEncoding = (archiveEncodingFieldInfo.GetValue(archive) as System.Text.Encoding?) ?? System.Text.Encoding.UTF8;
            
            foreach (ZipArchiveEntry entry in archive.Entries)
```

File: src/libraries/System.IO.Packaging/tests/ReflectionTests.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                FieldInfo fieldInfo = typeof(ZipArchiveEntry).GetField("_generalPurposeBitFlag", BindingFlags.Instance | BindingFlags.NonPublic);
+                object fieldObject = fieldInfo.GetValue(entry);
+                ushort shortField = (ushort)fieldObject;
+                Assert.Equal(0, shortField & 0x800); // If it was UTF8, we would set the general purpose bit flag to 0x800 (UnicodeFileNameAndComment)
+                CheckCharacters(entry.Name);
+                CheckCharacters(entry.Comment); // Unavailable in .NET Framework
```
Comment (edwardneal) [inline src/libraries/System.IO.Packaging/tests/ReflectionTests.cs:39]:
```suggestion
                fieldInfo = typeof(ZipArchiveEntry).GetField("_fileComment", BindingFlags.Instance | BindingFlags.NonPublic);
                byte[]? commentBytes = (byte[]?)fieldInfo.GetValue(archiveEntry);
                
                CheckCharacters(archiveEncoding.GetString(commentBytes)); // Unavailable in .NET Framework
```

**Review by edwardneal — CHANGES_REQUESTED** [review]
Thanks @carlossanlop. .NET 6.0 doesn't have the `ZipArchiveEntry.Comment` property, which complicates the backport slightly. I've added a change which should hopefully allow the test to keep on testing the specification behaviour regardless of this.

The simpler alternative is just to remove the reference to `entry.Comment` on line 39 of ReflectionTests.cs. It depends whether we need to test for behaviour when it's in the spec, but not in the public API surface.

File: src/libraries/System.IO.Packaging/tests/ReflectionTests.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                sw.Write(Tests.s_DocumentXml);
+            }
+            package.CreateRelationship(part.Uri, TargetMode.Internal, "http://packageRelType", "rId1234");
+        }
+
+        ms.Position = 0;
+        using (ZipArchive archive = new ZipArchive(ms, ZipArchiveMode.Read, leaveOpen: false))
+        {
+            FieldInfo archiveEncodingFieldInfo = typeof(ZipArchive).GetField("_entryNameAndCommentEncoding", BindingFlags.Instance | BindingFlags.NonPublic);
+            System.Text.Encoding archiveEncoding = (archiveEncodingFieldInfo.GetValue(archive) as System.Text.Encoding?) ?? System.Text.Encoding.UTF8;
```
Comment (carlossanlop) [inline src/libraries/System.IO.Packaging/tests/ReflectionTests.cs:33]:
```suggestion
            System.Text.Encoding archiveEncoding = (archiveEncodingFieldInfo.GetValue(archive) as System.Text.Encoding) ?? System.Text.Encoding.UTF8;
```

File: src/libraries/System.IO.Packaging/tests/ReflectionTests.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            
+            foreach (ZipArchiveEntry entry in archive.Entries)
+            {
+                FieldInfo fieldInfo = typeof(ZipArchiveEntry).GetField("_generalPurposeBitFlag", BindingFlags.Instance | BindingFlags.NonPublic);
+                object fieldObject = fieldInfo.GetValue(entry);
+                ushort shortField = (ushort)fieldObject;
+                Assert.Equal(0, shortField & 0x800); // If it was UTF8, we would set the general purpose bit flag to 0x800 (UnicodeFileNameAndComment)
+                CheckCharacters(entry.Name);
+                fieldInfo = typeof(ZipArchiveEntry).GetField("_fileComment", BindingFlags.Instance | BindingFlags.NonPublic);
+                byte[]? commentBytes = (byte[]?)fieldInfo.GetValue(archiveEntry);
```
Comment (carlossanlop) [inline src/libraries/System.IO.Packaging/tests/ReflectionTests.cs:43]:
```suggestion
                byte[]? commentBytes = (byte[]?)fieldInfo.GetValue(entry);
```

**Comment by edwardneal** [comment]
Thanks @carlossanlop - I see what you mean about the variable names and test errors. Apologies for those!

The test errors appear to be because the general purpose bit flag has the Unicode encoding bit (bit 11) set in the general purpose bit flags. The test is simply to make sure the compression levels roundtrip, so arguably this could be masked away and the test would pass. I'll check the tests over the next few days and submit another review with corrections.

File: src/libraries/System.IO.Packaging/tests/ReflectionTests.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                using StreamWriter sw = new(partStream);
+                sw.Write(Tests.s_DocumentXml);
+            }
+            package.CreateRelationship(part.Uri, TargetMode.Internal, "http://packageRelType", "rId1234");
+        }
+
+        ms.Position = 0;
+        using (ZipArchive archive = new ZipArchive(ms, ZipArchiveMode.Read, leaveOpen: false))
+        {
+            FieldInfo archiveEncodingFieldInfo = typeof(ZipArchive).GetField("_entryNameAndCommentEncoding", BindingFlags.Instance | BindingFlags.NonPublic);
```
Comment (edwardneal) [inline src/libraries/System.IO.Packaging/tests/ReflectionTests.cs:32]:
```suggestion
            FieldInfo archiveEncodingFieldInfo = typeof(ZipArchive).GetField("_entryNameEncoding", BindingFlags.Instance | BindingFlags.NonPublic);
```

**Review by edwardneal — COMMENTED** [review]
Finding the cause took less time than I thought.

By default, .NET 6.0 creates the ZipArchive with UTF8 encoding, which eventually feeds through to setting bit 11 in the general purpose bitflag for each archive entry. PR #87883 corrects this for .NET 8.0, and hasn't been backported to .NET 6.0 - so the values of the general purpose bit flags for a ZIP archive entry will be different between .NET 6.0 and .NET 8.0.

The purpose of the `Roundtrip_Compression_Option` test only requires us to validate bits 1 & 2, so I've just decided to mask the other bits out. The exception should never be triggered in the .NET 8.0 test because we don't create any files which need the extended encoding support, but I can update it to keep it in sync with the .NET 6.0 test if needed.

The other two review comments are just the result of me checking the tests in Visual Studio rather than eyeballing the code in GitHub...

**Comment by carlossanlop** [comment]
@ViktorHofer @ericstj can I get a review and a sign-off for this servicing fix, please?

**Comment by carlossanlop** [comment]
Closing while we get clarity on whether this bug's impact is severe enough to justify a backport.

**Pull request closed without merging**